### PR TITLE
Add plugin adapters.

### PR DESCRIPTION
**Pull request by Cybermaxke**
**API** | [Common](https://github.com/SpongePowered/SpongeCommon/pull/2238) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/2694) | Vanilla | [KotlinAdapter](https://github.com/Cybermaxke/KotlinAdapter)

Supports alternate plugin instantiation and injections. Can be used e.g. for kotlin/scala object support, kotlin delegate property injections, custom plugin injections, etc.

I am going to write a test project to demonstrate how it can be used. SpongeVanilla support will follow once it's successfully tested on SpongeForge.

Kotlin Adapter:
https://github.com/Cybermaxke/KotlinAdapter/blob/master/src/main/kotlin/org/lanternpowered/kt/adapter/KotlinAdapter.kt

Kotlin Test Plugin:
https://github.com/Cybermaxke/KotlinAdapter/blob/master/src/main/kotlin/org/lanternpowered/kt/plugin/TestPlugin.kt

### CODE REVIEW

**Comment by XakepSDK** [comment]
Can you also add support for guice modules for plugins?
E.g. developer specifies module class in plugin annotation and then impl. installs that module for plugin's injector.

File: src/main/java/org/spongepowered/api/plugin/PluginAdapter.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * @param defaultInjector The default injector
+     * @param pluginModules A immutable list of modules which are provided by the plugin
+     * @return The injector
+     */
+    <T> Injector getInjector(PluginContainer pluginContainer, Class<T> pluginClass, Injector defaultInjector, List<Module> pluginModules);
+
+    /**
+     * Represents the default {@link PluginAdapter}.
+     */
+    final class Default implements PluginAdapter {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/plugin/PluginAdapter.java]: Should not be implemented here - should be an empty interface, and implemented elsewhere.

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/plugin/PluginAdapter.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+     * <p>Calling {@link PluginContainer#getInstance()} during this
+     * method will result in a {@link IllegalStateException}.</p>
+     *
+     * @param pluginContainer The plugin container the injector is being created for
+     * @param pluginClass The plugin class
+     * @param defaultInjector The default injector
+     * @param pluginModules A immutable list of modules which are provided by the plugin
+     * @return The injector
+     */
+    <T> Injector getInjector(PluginContainer pluginContainer, Class<T> pluginClass, Injector defaultInjector, List<Module> pluginModules);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/plugin/PluginAdapter.java]: `createInjector`

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by mattmess1221** [comment]
👎 for cluttering up the `@Plugin` class.

If we go the route that FML went, the annotation would be specific to the default java loader. The Kotlin loader would have to provide its own loading scheme.

I would also like to see this done as a [ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html).

**Comment by Cybermaxke** [comment]
I don't think it's cluttering the annotation, it's not like it's required which could clutter the usage. Same goes for the modules. We could always move them to separate annotations if needed.

Going down the road of reimplementing the complete loader is going to complicate things, you'll need to reimplement the `PluginContainer`, handle class loading, annotation scanning. And in the end, we'll just end up with yet another `Plugin` annotation. I also don't see how this complete control is going to add any more value than what's already possible by injector transformations.

The goal of the plugin adapters isn't to reimplement of what's already there, replacing the complete loading system isn't worth it IMO.

If you also look at what forge uses it for in there scala adapter, it's just to provide language specific injections. Which can easily be handled by guice.

**Comment by pie-flavor** [comment]
We'd need some way of instantiating PluginContainer in a safe way if we had separate annotations. This actually seems like *less* bloat to add another field to Plugin.

**Comment by Katrix** [comment]
Will there be any way to avoid having to interface with Guice so that I can wire up by stuff statically? I don't really see any value in this from a Scala point of view at least if I can't.

Also, a bit unsure what you're talking about with the Scala adapter for Forge injecting language specific stuff. It doesn't inject anything. No mod or plugin annotation, or anything like that. You specify the name of the plugin object in the toml file, and go on your way.

**Comment by Cybermaxke** [comment]
You don't have to use guice injections in your scala plugin, sponge uses it to construct the instance. You can register a provider and implement it however you like.

As for the scala adapter, I looked at a older version that didn't use any config, when it was still in Forge.

I also don't feel like exposing the construction of `PluginContainer`s is a good idea, maybe a way to register a alternate annotation and map those to PluginContainer values is better? And optionally registering adapters, etc.
E.g. 
```kotlin
annotation class KotlinPlugin(val id: String, val name: String)

// Find a way to register annotations?
registry.registerPluginAnnotation(KotlinPlugin::class.java, KotlinPlugin::id).mapName(KotlinPlugin::name).adapter(MyAdapter())
```

**Comment by pie-flavor** [comment]
Actually it would be nice to not have to use the Plugin annotation, or any annotation at all. This API could then be extended to stuff not even Java-based, such as scripting plugins, so they could register each script set as a plugin.

**Comment by dualspiral** [comment]
Are we going to revisit this for API 7, or are we just going to leave this for the [Launcher](https://github.com/SpongePowered/plugin-spi) project in API 8?

**Comment by Zidane** [comment]
I am closing this.

As dual linked, plugin-spi is very lightweight and super simple to implement a language loading service for other JVM based languages (or even ones that aren't anchored in the JVM). You can easily add kotlin/scala features to your own service.

No plans for api-7 to gain this change.

**Pull request closed without merging**